### PR TITLE
12311 Disable Nagle in HTTP client and server

### LINKED ISSUE #12311: Disable Nagle in HTTP client and server
After spending a bunch of time trying to figure out why Tahoe-LAFS is spending a surprising amount of time idle while doing HTTP interactions, I discovered it was Nagle adding 40ms to small queries. 

The argument for disabling Nagle / enabling TCP_NODELAY:

* HTTP is often used for small, RPC-like queries, so delaying queries by 40ms is bad.
* Large bodies typically won't benefit Nagle anyway.
* HTTPS is all anyone uses in practice these days, and Twisted's TLS layer already does some minimal buffering to reduce small writes: https://github.com/twisted/twisted/blob/f0afc5eb9850eda242beefb178dd481e28926f08/src/twisted/protocols/tls.py#L818

### PR DESCRIPTION
**Pull request by itamarst**
## Scope and purpose

Fixes #12311 

Setting TCP_NODELAY will result in reduced latency for small requests, instead of waiting e.g. 40ms before sending the request.

### CODE REVIEW

**Comment by itamarst** [comment]
Blocked by #12314, should pass once that is merged.

**Comment by itamarst** [comment]
I don't understand how to prevent the test failure; locally everything works fine and `test_http2` get skipped as expected.

File: pyproject.toml
Comment (itamarst) [inline pyproject.toml:99]: Necessary to keep httpx happy in the tests. 3.2.0 is from February 2020, still plenty of leeway.

File: src/twisted/web/test/test_http2.py
Comment (itamarst) [inline src/twisted/web/test/test_http2.py]: Necessary to ensure cover for the change in `_http2.py`.

File: src/twisted/web/_newclient.py
Comment (glyph) [inline src/twisted/web/_newclient.py:1501]: I very much think the answer is "no" but since I noticed this I should probably raise it: do we have to care about inheritance here, and calling `super().connectionMade()`? Hopefully this is just private and we totally don't have to care

File: src/twisted/web/_http2.py
Comment (glyph) [inline src/twisted/web/_http2.py:147]: hooray we definitely don't have to care about super() as I fret about below because we already overrode this method

File: src/twisted/web/test/test_http2.py
Comment (glyph) [inline src/twisted/web/test/test_http2.py]:
```suggestion
    async def test_realRequest(self) -> None:
```

Comment (glyph) [reply src/twisted/web/test/test_http2.py]: let's not add more untyped code if we can help it

Comment (glyph) [reply src/twisted/web/test/test_http2.py]: Neat, I assume this coroutine is running because it's covered, when did we add that to trial?

**Review by glyph — APPROVED** [review]
modulo the hopefully-dismissable concern about inheritance, lgtm

File: src/twisted/web/test/test_http2.py
Comment (itamarst) [reply src/twisted/web/test/test_http2.py]: Not sure but I switched to this style of async test in Tahoe a while ago.

File: src/twisted/web/_newclient.py
Comment (itamarst) [reply src/twisted/web/_newclient.py:1501]: It's a private class, so inheritance isn't an issue. And `LineReceiver` has no connectionMade so I think it's fine.

**Comment by p12tic** [comment]
@itamarst @glyph 

This PR is likely to break connections made over `unix:` protocol. See https://github.com/buildbot/buildbot/issues/8207.

Relevant log:
```
2024-11-12 05:54:43+0000 [buildbot.www.service.BuildbotSite] Unhandled Error
        Traceback (most recent call last):
          File "/home/buildbot/venv/lib64/python3.13/site-packages/twisted/internet/tcp.py", line 1423, in doRead
            protocol.makeConnection(transport)
          File "/home/buildbot/venv/lib64/python3.13/site-packages/twisted/internet/protocol.py", line 509, in makeConnection
            self.connectionMade()
          File "/home/buildbot/venv/lib64/python3.13/site-packages/twisted/web/http.py", line 2336, in connectionMade
            self.transport.setTcpNoDelay(True)
          File "/home/buildbot/venv/lib64/python3.13/site-packages/twisted/internet/tcp.py", line 340, in setTcpNoDelay
            self.socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, enabled)
        builtins.OSError: [Errno 95] Operation not supported
```

Relevant Twisted configuration:
```
    port='unix:/run/buildbot/buildbot-loop-master.sock:mode=660',
```

I haven't done checks yet to prove that this PR is indeed to blame.

**Pull request merged**
